phpBB/phpbb/mention/helper/mention_helper.php

```diff
@@ -0,0 +1,123 @@
+		$this->data = $data;
+		$regular_expression_match = '#\[mention\](.*?)\[/mention\]#';
+		$matches = false;
+		$user_list = $this->get_userlist();
+		preg_match_all($regular_expression_match, $post_text, $matches, PREG_OFFSET_CAPTURE);
+		$start_tag_length = strlen("[mention]");
+		$end_tag_length = strlen("[\mention]");
+		$users_notified = array();
+		for ($i = 0, $len = count($matches[1]); $i < $len; $i++)
+		{
+			$username_clean = utf8_clean_string($matches[1][$i][0]);
+			if (array_key_exists($username_clean, $user_list))
+			{
+				if(!in_array($username_clean, $users_notified, true)){
+
+					$startpos = $matches[1][$i][1] - $start_tag_length;
+					$length = strlen($matches[1][$i][0]);
+					$endpos = $matches[1][$i][1] + $length + $end_tag_length - 1;
+					$user_data = $user_list[$username_clean];
+					$add_url_tag = "[url= memberlist.php?mode=viewprofile&u=".$user_data."]".$username_clean."[/url]";
+					if ($i==0) {
+						$new_post_text = substr($post_text, 0, $startpos);
+						$new_post_text = $new_post_text . $add_url_tag;
+						$prev_end_pos = $endpos + 1;
+					}else {
+						$strip_mention = substr($post_text, $prev_end_pos, $startpos - $prev_end_pos);
+						$new_post_text = $new_post_text . $strip_mention;
+						$new_post_text = $new_post_text . $add_url_tag;
+						$prev_end_pos = $endpos + 1;
+					}
+					$this->send_notifications($user_list[$username_clean], $notif_manager_obj);
+				}
+				array_push($users_notified, $username_clean);
+			}
+		}
+		return $new_post_text;
+
+	}
+
+	public function get_userlist() {
+		$sql_ary = "SELECT username, user_id, username_clean FROM ". USERS_TABLE;
+		$result = $this->db->sql_query($sql_ary);
+		$user_list = array();
+		while ($row = $this->db->sql_fetchrow($result))
+		{
+			// print_r($row);
+			$user_list[$row['username_clean']] = $row["user_id"];
+		}
+		$this->db->sql_freeresult($result);
+		return $user_list;
+	}
+
+	public function send_notifications($user_id, $notif_manager_obj){
+		$notification_method = array();
+		$notification_details_list = $this->get_notification_type_and_method($user_id);
+		// echo print_r($notification_details_list);
+		foreach($notification_details_list as $details) {
+			$notification_type = $notif_manager_obj->get_item_type_class($details["notif_type"]);
+			$notification_type->user_id = $user_id;
+			$notification_type->create_insert_array($this->data);
+
+			$notification_method[$details["notif_method"]] = $notif_manager_obj->get_method_class($details["notif_method"]);
+			$notification_method[$details["notif_method"]]->add_to_queue($notification_type);
+		}
+		// echo sizeof($notification_method);
+		foreach($notification_method as $method){
+			// echo get_class($method);
+			$method->notify();
+		}
+	}
+
+	public function get_notification_type_and_method($user_id){
+
+		$sql_ary = "SELECT item_type, method from phpbb_user_notifications where user_id =$user_id";
```

Comment:
You should inject the table name into your class, if a user is not using the default prefix this will not work.

Also, Please cast $user_id to a integer.